lib/spack/spack/test/compilers/detection.py

```diff
@@ -24,6 +25,51 @@
 import spack.compilers.xl_r
 import spack.util.module_cmd
 from spack.operating_systems.cray_frontend import CrayFrontend
+from spack.pkg.builtin.aocc import Aocc
+from spack.pkg.builtin.apple_clang import AppleClang
+from spack.pkg.builtin.arm import Arm
+from spack.pkg.builtin.cce import Cce
+from spack.pkg.builtin.fj import Fj
+from spack.pkg.builtin.gcc import Gcc
+from spack.pkg.builtin.intel_oneapi_compilers_classic import IntelOneapiCompilersClassic
+from spack.pkg.builtin.intel_oneapi_compilers import IntelOneapiCompilers
+from spack.pkg.builtin.nag import Nag
+from spack.pkg.builtin.nvhpc import Nvhpc
+from spack.pkg.builtin.pgi import Pgi
+from spack.pkg.builtin.xl import Xl
+
+
+def check_package_detection(mock_executable, output, expected_version, cls):
+    languages = cls.compiler_languages
+
+    executables = {}
+    script = "\n".join(f'echo "{line}"' for line in output.split("\n"))
+    for lang in languages:
+        name = getattr(cls, f"{lang}_names")[0]
+        executables[name] = mock_executable(name, output=script)
+        bindir = os.path.dirname(executables[name])
+        with open(executables[name], "r") as f:
+            print(f.read())
+
+    detected = spack.detection.by_path(
+        [cls.fullname.replace("_", "-")], path_hints=[bindir], max_workers=1
+    )
+
+    spec_name = cls.name.replace("_", "-")
+    assert len(detected) == 1
+    assert spec_name in detected
+
+    detected_package = detected[spec_name]
+    assert len(detected_package) == 1
+    assert detected_package[0].spec.satisfies(f"{spec_name}@{expected_version}")
+    assert detected_package[0].prefix == os.path.dirname(bindir)
+
+    suffix = ".bat" if sys.platform == "win32" else ""
+    for lang in languages:
+        name = getattr(cls, f"{lang}_names")[0]
+        assert detected_package[0].spec.extra_attributes["paths"][lang] == os.path.join(
+            bindir, name + suffix
+        )
```

Comment:
Why are we adding this? We already have `detection_tests.yaml` to check package detection. I don't get why we are duplicating the test...

Reply:
`detection_tests.yaml` runs on a single architecture. I've already caught bugs in this PR that wouldn't have been caught by `detection_tests.yaml`.

Reply:
Which bugs? Why they would not be caught?

Reply:
Windows-specific bugs -- we don't run audits on Windows

Reply:
> Windows-specific bugs -- we don't run audits on Windows.

Why not adding audits on Windows then? 

> Also some bugs that could have been expressed in an audit [...]

I'm not sure I understand what you mean here.

Reply:
Windows audits have been added, and the ability to check extra attributes too. I'll try to remove this extra code.

Reply:
Extra code removed.